lib/Frontend/CompilerInvocation.cpp

```diff
@@ -2460,27 +2460,91 @@ static bool ParseSearchPathArgs(SearchPathOptions &Opts, ArgList &Args,
   return false;
 }
 
+static bool
+ParseDiagnosticVerifierArgs(std::optional<DiagnosticVerifierOptions> &Opts,
+                            ArgList &Args, DiagnosticEngine &Diags,
+                            llvm::opt::OptTable &Table) {
+  using namespace options;
+
+  const auto verifyArgs = Args.filtered(OPT_verify);
+  const bool applyFixes = Args.hasArg(OPT_verify_apply_fixes);
+  const bool ignoreUnknown = Args.hasArg(OPT_verify_ignore_unknown);
+
+  auto additionalFilePathArgs = Args.filtered(OPT_verify_additional_file);
+  auto additionalPrefixArgs = Args.filtered(OPT_verify_additional_prefix);
+
+  if (verifyArgs.empty() && !applyFixes) {
+    if (ignoreUnknown) {
+      Diags.diagnose(
+          SourceLoc(), diag::ignoring_option_requires_option,
+          Table.getOption(OPT_verify_ignore_unknown).getPrefixedName(),
+          Table.getOption(OPT_verify).getPrefixedName());
+    }
+    if (!additionalFilePathArgs.empty()) {
+      Diags.diagnose(
+          SourceLoc(), diag::ignoring_option_requires_option,
+          Table.getOption(OPT_verify_additional_file).getPrefixedName(),
+          Table.getOption(OPT_verify).getPrefixedName());
+    }
+    if (!additionalPrefixArgs.empty()) {
+      Diags.diagnose(
+          SourceLoc(), diag::ignoring_option_requires_option,
+          Table.getOption(OPT_verify_additional_prefix).getPrefixedName(),
+          Table.getOption(OPT_verify).getPrefixedName());
+    }
+
+    return false;
+  }
+
+  DiagnosticVerifierOptions verifierOpts;
+
+  verifierOpts.ApplyFixes = applyFixes;
+  verifierOpts.IgnoreUnknown = ignoreUnknown;
+
+  for (auto *arg : verifyArgs) {
+    auto &values = arg->getValues();
+    if (values.empty()) {
+      continue;
+    }
+
+    for (StringRef value : values) {
+      if (value == "ignore-unknown") {
+        verifierOpts.IgnoreUnknown = true;
+      } else if (value == "apply-fixes") {
+        verifierOpts.ApplyFixes = true;
+      } else if (value == "no-errors") {
+        verifierOpts.NoErrors = true;
+      } else {
+        Diags.diagnose(SourceLoc(), diag::error_unsupported_option_argument,
```

Comment:
Don't we need to propagate the fact that an error diagnostic was emitted here by returning `true` to ensure compiler exits properly?

Reply:
We do indeed, apparently. The exit code is zero, which is somewhat surprising. I was expecting the frontend to ask the engine if any error were emitted before exiting, but it seems to rely on the return value of the various `perform*` functions.

Reply:
This is the sort of situation where I want a non-zero exit code but don’t want the error to abort compilation.